<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Example Workflow: CELLxGENE</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Example Workflow: CELLxGENE</h1>



<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>This vignette demonstrates a basic workflow for accessing and
analysing single-cell RNA-seq data from the CELLxGENE repository using
<strong>{laminr}</strong>. <a href="https://cellxgene.cziscience.com/">CZ CELLxGENE Discover</a> is a
standardised collection of scRNA-seq datasets and LaminDB makes it easy
to query and access data in this repository. We will go through the
steps of finding and downloading a dataset using
<strong>{laminr}</strong>, performing some simple analysis using
<strong>{Seurat}</strong> and saving the results your own LaminDB
database.</p>
</div>
<div id="before-we-start" class="section level1">
<h1>Before we start</h1>
<p>Before we go begin, please take some time to check out the Getting
Started vignette (<code>vignette(&quot;laminr&quot;, package = &quot;laminr&quot;)</code>).
In particular, make sure you have run the commands in the “Initial
Setup” section.</p>
<p>Once that is done, we can load the <strong>{laminr}</strong>
library.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(laminr)</span></code></pre></div>
</div>
<div id="connecting-to-lamindb" class="section level1">
<h1>Connecting to LaminDB</h1>
<p>The first thing we need to do is connect to the LaminDB database. For
this tutorial, we will connect a default instance (where we will store
results) and the CELLxGENE instance that we will search for
datasets.</p>
<div id="connect-to-the-default-instance" class="section level2">
<h2>Connect to the default instance</h2>
<p>We will start by connecting to your default LaminDB instance. You can
set set the default instance using the <code>lamin</code> CLI on the
command line:</p>
<pre class="shell"><code>lamin connect &lt;owner&gt;/&lt;name&gt;</code></pre>
<p>Once a default instance has been set, we can connect to it with
<strong>{laminr}</strong>:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">connect</span>()</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="co">#&gt; [92m→[0m connected lamindb: laminlabs/lamindata</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>db</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="co">#&gt; lamindata</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="co">#&gt;   Core registries</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="co">#&gt;     $Run</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="co">#&gt;     $User</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="co">#&gt;     $Param</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="co">#&gt;     $ULabel</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="co">#&gt;     $Feature</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a><span class="co">#&gt;     $Storage</span></span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a><span class="co">#&gt;     $Artifact</span></span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a><span class="co">#&gt;     $Transform</span></span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a><span class="co">#&gt;     $Collection</span></span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a><span class="co">#&gt;     $FeatureSet</span></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a><span class="co">#&gt;     $ParamValue</span></span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a><span class="co">#&gt;     $FeatureValue</span></span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a><span class="co">#&gt;   Additional modules</span></span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a><span class="co">#&gt;     bionty</span></span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a><span class="co">#&gt;     wetlab</span></span></code></pre></div>
<p>This gives us an object we can use to interact with the database.</p>
<p><strong>Note</strong> that only the default instance can create new
records. This tutorial assumes you have access to an instance where you
have permission to add data.</p>
</div>
<div id="track-data-provenance" class="section level2">
<h2>Track data provenance</h2>
<p>Before we start, we will track the code that is run in this
notebook.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>db<span class="sc">$</span><span class="fu">track</span>(<span class="st">&quot;I8BlHXFXqZOG0000&quot;</span>, <span class="at">path =</span> <span class="st">&quot;example_workflow.Rmd&quot;</span>)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="co">#&gt; [92m→[0m created Transform(&#39;I8BlHXFX&#39;), started new Run(&#39;CKjEjAr6&#39;) at 2024-11-20 11:17:56 UTC</span></span></code></pre></div>
<p>Tip: The ID should be obtained by running
<code>db$track(path = &quot;example_workflow.Rmd&quot;)</code> and copying the ID
from the output.</p>
</div>
<div id="connect-to-the-cellxgene-instance" class="section level2">
<h2>Connect to the CELLxGENE instance</h2>
<p>We can connect to other instances by providing a slug to the
<code>connect()</code> function. Instances connected to in this way can
be used to query data but cannot make any changes. Let’s connect to the
CELLxGENE instance:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>cellxgene <span class="ot">&lt;-</span> <span class="fu">connect</span>(<span class="st">&quot;laminlabs/cellxgene&quot;</span>)</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>cellxgene</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="co">#&gt; cellxgene</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="co">#&gt;   Core registries</span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="co">#&gt;     $Run</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="co">#&gt;     $User</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="co">#&gt;     $Param</span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="co">#&gt;     $ULabel</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="co">#&gt;     $Feature</span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="co">#&gt;     $Storage</span></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="co">#&gt;     $Artifact</span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="co">#&gt;     $Transform</span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a><span class="co">#&gt;     $Collection</span></span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a><span class="co">#&gt;     $FeatureSet</span></span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a><span class="co">#&gt;     $ParamValue</span></span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a><span class="co">#&gt;     $FeatureValue</span></span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a><span class="co">#&gt;   Additional modules</span></span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a><span class="co">#&gt;     bionty</span></span></code></pre></div>
</div>
</div>
<div id="downloading-a-dataset" class="section level1">
<h1>Downloading a dataset</h1>
<p>In Lamin, artifacts are objects that contain information (single-cell
data, images, data frames etc.) as well as associated metadata. You can
see what artifacts are available using the database instance object.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>cellxgene<span class="sc">$</span>Artifact<span class="sc">$</span><span class="fu">df</span>(<span class="at">limit =</span> <span class="dv">5</span>)</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="co">#&gt;     id suffix X_accessor n_objects visibility                                                                    key                  uid         size                   hash</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="co">#&gt; 1 2846        tiledbsoma       290          1                                            cell-census/2023-12-15/soma FYMewVq5twKMDXVy0000 635848093433 Mfyw8VuqftX5REITfQH_yg</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="co">#&gt; 2 3665        tiledbsoma       330          1                                            cell-census/2024-07-01/soma FYMewVq5twKMDXVy0001 870700998221 bzrXBPNvitSVKvb3GG38_w</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="co">#&gt; 3 1270  .h5ad    AnnData        NA          1 cell-census/2023-07-25/h5ads/7a0a8891-9a22-4549-a55b-c2aca23c3a2a.h5ad tczTlSHFPOcAcBnfyxKA   1297573950 UlsVvBz9kMzn2r9RdoAAOg</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="co">#&gt; 4 2840 .ipynb       &lt;NA&gt;        NA          0                                                                   &lt;NA&gt; JIIPyQX5l9qELPl42d75        36297 gNdUkonYgQJP_Mi3xLzt_g</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="co">#&gt; 5 2842  .html       &lt;NA&gt;        NA          0                                                                   &lt;NA&gt; Whyxwf3k2GjJwTPCl1FK       716529 BDGZac3qU3oLVFpO035Qhg</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="co">#&gt;                            description n_observations is_latest X_hash_type    type                       created_at X_key_is_virtual                       updated_at    version</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="co">#&gt; 1                    Census 2023-12-15       68683222     FALSE       md5-d dataset 2024-07-12T12:12:16.091881+00:00            FALSE 2024-09-17T13:00:13.714256+00:00 2023-12-15</span></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="co">#&gt; 2                    Census 2024-07-01      115556140      TRUE       md5-d dataset 2024-07-16T12:52:01.424629+00:00            FALSE 2024-09-17T13:01:23.739635+00:00 2024-07-01</span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="co">#&gt; 3      Supercluster: Hippocampal CA1-3          74979     FALSE       md5-n    &lt;NA&gt; 2023-11-28T21:46:12.685907+00:00            FALSE 2024-01-24T07:10:21.725547+00:00 2023-07-25</span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a><span class="co">#&gt; 4 Source of transform G69jtgzKO0eJ6K79             NA     FALSE         md5    &lt;NA&gt; 2024-01-29T08:32:13.311741+00:00             TRUE 2024-01-29T08:32:13.311792+00:00          0</span></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a><span class="co">#&gt; 5   Report of run UAAiLAi0BrLvlKnsuvP3             NA     FALSE         md5    &lt;NA&gt; 2024-01-29T08:32:18.346499+00:00             TRUE 2024-01-30T09:12:06.027928+00:00          1</span></span></code></pre></div>
<p>This is useful, but it’s not the nicest or easiest way to find a
particular dataset. Instead, we will use the Lamin Hub website to find
the data we want to load.</p>
<ol style="list-style-type: decimal">
<li>Open a browser and go to <a href="https://lamin.ai/laminlabs/cellxgene" class="uri">https://lamin.ai/laminlabs/cellxgene</a></li>
<li>On the top toolbar, click the “Artifacts” tab</li>
<li>Use the search field and the filters to find a dataset you are
interested in.</li>
</ol>
<ul>
<li>We use the “Suffix” filter to find <code>.h5ad</code> files and
search for “renal cell carcinoma”</li>
</ul>
<ol start="4" style="list-style-type: decimal">
<li>Select the entry for the dataset you want to load to open a page
with more details</li>
<li>Click the copy button at the top right, this copies a command
including the ID for the artifact</li>
</ol>
<p>Once we have the artifact ID, we can load information about the
artifact, similar to what we see on the website. Notice that we use a
slightly different command to what we copied from the website.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>artifact <span class="ot">&lt;-</span> cellxgene<span class="sc">$</span>Artifact<span class="sc">$</span><span class="fu">get</span>(<span class="st">&quot;7dVluLROpalzEh8mNyxk&quot;</span>)</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>artifact</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="co">#&gt; Artifact(uid=&#39;7dVluLROpalzEh8mNyxk&#39;, description=&#39;Renal cell carcinoma, pre aPD1, kidney Puck_200727_12&#39;, key=&#39;cell-census/2023-12-15/h5ads/02faf712-92d4-4589-bec7-13105059cf86.h5ad&#39;, id=1742, run_id=22, hash=&#39;YNYuokfAoDFxdaRILjmU9w&#39;, size=13997860, suffix=&#39;.h5ad&#39;, storage_id=2, version=&#39;2023-12-15&#39;, _accessor=&#39;AnnData&#39;, is_latest=TRUE, transform_id=16, _hash_type=&#39;md5-n&#39;, created_at=&#39;2024-01-11T09:13:23.143694+00:00&#39;, created_by_id=1, updated_at=&#39;2024-01-24T07:17:47.009288+00:00&#39;, visibility=1, n_observations=17612, _key_is_virtual=FALSE)</span></span></code></pre></div>
<p>So far we have only retrieved the metadata about this object. To
download the data itself we need to run another command.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>adata <span class="ot">&lt;-</span> artifact<span class="sc">$</span><span class="fu">load</span>()</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="co">#&gt; ℹ &#39;s3://cellxgene-data-public/cell-census/2023-12-15/h5ads/02faf712-92d4-4589-bec7-13105059cf86.h5ad&#39; already exists at &#39;/home/rcannood/.cache/lamindb/cellxgene-data-public/cell-census/2023-12-15/h5ads/02faf712-92d4-4589-bec7-13105059cf86.h5ad&#39;</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>adata</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="co">#&gt; AnnData object with n_obs × n_vars = 17612 × 23254</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="co">#&gt;     obs: &#39;n_genes&#39;, &#39;n_UMIs&#39;, &#39;log10_n_UMIs&#39;, &#39;log10_n_genes&#39;, &#39;Cell_Type&#39;, &#39;cell_type_ontology_term_id&#39;, &#39;organism_ontology_term_id&#39;, &#39;tissue_ontology_term_id&#39;, &#39;assay_ontology_term_id&#39;, &#39;disease_ontology_term_id&#39;, &#39;self_reported_ethnicity_ontology_term_id&#39;, &#39;development_stage_ontology_term_id&#39;, &#39;sex_ontology_term_id&#39;, &#39;donor_id&#39;, &#39;is_primary_data&#39;, &#39;suspension_type&#39;, &#39;cell_type&#39;, &#39;assay&#39;, &#39;disease&#39;, &#39;organism&#39;, &#39;sex&#39;, &#39;tissue&#39;, &#39;self_reported_ethnicity&#39;, &#39;development_stage&#39;</span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="co">#&gt;     var: &#39;gene&#39;, &#39;n_beads&#39;, &#39;n_UMIs&#39;, &#39;feature_is_filtered&#39;, &#39;feature_name&#39;, &#39;feature_reference&#39;, &#39;feature_biotype&#39;</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="co">#&gt;     uns: &#39;Cell_Type_colors&#39;, &#39;schema_version&#39;, &#39;title&#39;</span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="co">#&gt;     obsm: &#39;X_spatial&#39;</span></span></code></pre></div>
<p>This dataset has been stored as an <a href="https://anndata.readthedocs.io"><code>AnnData</code></a> object.
In the next sections we will convert it to a <a href="https://satijalab.org/seurat/"><code>Seurat</code></a> object and
perform some simple analysis.</p>
</div>
<div id="convert-to-seurat" class="section level1">
<h1>Convert to Seurat</h1>
<p>There are various approaches for converting between different
single-cell objects, some of which are described in the <a href="https://www.sc-best-practices.org/introduction/interoperability.html">Interoperability
chapter</a> of the Single-cell Best Practices book.</p>
<p>Because we already have the data loaded in memory, the simplest
option is to extract the information we need and create a new
<code>Seurat</code> object.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>seurat <span class="ot">&lt;-</span> SeuratObject<span class="sc">::</span><span class="fu">CreateSeuratObject</span>(</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>  <span class="at">counts =</span> Matrix<span class="sc">::</span><span class="fu">t</span>(adata<span class="sc">$</span>X),</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>  <span class="at">meta.data =</span> adata<span class="sc">$</span>obs,</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>)</span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="co">#&gt; Warning: Data is of class dgRMatrix. Coercing to dgCMatrix.</span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>seurat</span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a><span class="co">#&gt; An object of class Seurat </span></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a><span class="co">#&gt; 23254 features across 17612 samples within 1 assay </span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a><span class="co">#&gt; Active assay: RNA (23254 features, 0 variable features)</span></span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a><span class="co">#&gt;  1 layer present: counts</span></span></code></pre></div>
</div>
<div id="analysis" class="section level1">
<h1>Analysis</h1>
<p>We could perform any normal analysis using <strong>{Seurat}</strong>
but as an example we will calculate marker genes for each of the
annotated cell types. To make things a bit quicker we only test the
first 1000 genes but if you have a few minutes you can get results for
all features.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co"># Set cell identities to the provided cell type annotation</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>SeuratObject<span class="sc">::</span><span class="fu">Idents</span>(seurat) <span class="ot">&lt;-</span> <span class="st">&quot;Cell_Type&quot;</span></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="co"># Normalise the data</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>seurat <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">NormalizeData</span>(seurat)</span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="co">#&gt; Normalizing layer: counts</span></span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a><span class="co"># Test for marker genes</span></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a>markers <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">FindAllMarkers</span>(</span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a>  seurat,</span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a>  <span class="at">features =</span> SeuratObject<span class="sc">::</span><span class="fu">Features</span>(seurat)[<span class="dv">1</span><span class="sc">:</span><span class="dv">1000</span>]</span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a>)</span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a><span class="co">#&gt; Calculating cluster Epithelial</span></span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a><span class="co">#&gt; Calculating cluster Fibroblast</span></span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a><span class="co">#&gt; For a (much!) faster implementation of the Wilcoxon Rank Sum Test,</span></span>
<span id="cb10-14"><a href="#cb10-14" tabindex="-1"></a><span class="co">#&gt; (default method for FindMarkers) please install the presto package</span></span>
<span id="cb10-15"><a href="#cb10-15" tabindex="-1"></a><span class="co">#&gt; --------------------------------------------</span></span>
<span id="cb10-16"><a href="#cb10-16" tabindex="-1"></a><span class="co">#&gt; install.packages(&#39;devtools&#39;)</span></span>
<span id="cb10-17"><a href="#cb10-17" tabindex="-1"></a><span class="co">#&gt; devtools::install_github(&#39;immunogenomics/presto&#39;)</span></span>
<span id="cb10-18"><a href="#cb10-18" tabindex="-1"></a><span class="co">#&gt; --------------------------------------------</span></span>
<span id="cb10-19"><a href="#cb10-19" tabindex="-1"></a><span class="co">#&gt; After installation of presto, Seurat will automatically use the more </span></span>
<span id="cb10-20"><a href="#cb10-20" tabindex="-1"></a><span class="co">#&gt; efficient implementation (no further action necessary).</span></span>
<span id="cb10-21"><a href="#cb10-21" tabindex="-1"></a><span class="co">#&gt; This message will be shown once per session</span></span>
<span id="cb10-22"><a href="#cb10-22" tabindex="-1"></a><span class="co">#&gt; Calculating cluster Myeloid</span></span>
<span id="cb10-23"><a href="#cb10-23" tabindex="-1"></a><span class="co">#&gt; Calculating cluster Tumor</span></span>
<span id="cb10-24"><a href="#cb10-24" tabindex="-1"></a><span class="co">#&gt; Warning: The following tests were not performed:</span></span>
<span id="cb10-25"><a href="#cb10-25" tabindex="-1"></a><span class="co">#&gt; Warning: When testing Epithelial versus all:</span></span>
<span id="cb10-26"><a href="#cb10-26" tabindex="-1"></a><span class="co">#&gt;  Cell group 1 has fewer than 3 cells</span></span>
<span id="cb10-27"><a href="#cb10-27" tabindex="-1"></a><span class="co"># The output is a data.frame</span></span>
<span id="cb10-28"><a href="#cb10-28" tabindex="-1"></a><span class="fu">head</span>(markers)</span>
<span id="cb10-29"><a href="#cb10-29" tabindex="-1"></a><span class="co">#&gt;                        p_val avg_log2FC pct.1 pct.2    p_val_adj    cluster            gene</span></span>
<span id="cb10-30"><a href="#cb10-30" tabindex="-1"></a><span class="co">#&gt; ENSG00000164283 1.030703e-89  2.7485040 0.205 0.048 2.396797e-85 Fibroblast ENSG00000164283</span></span>
<span id="cb10-31"><a href="#cb10-31" tabindex="-1"></a><span class="co">#&gt; ENSG00000116016 3.606838e-38  2.0721038 0.152 0.051 8.387340e-34 Fibroblast ENSG00000116016</span></span>
<span id="cb10-32"><a href="#cb10-32" tabindex="-1"></a><span class="co">#&gt; ENSG00000074800 5.097282e-25 -0.9810317 0.185 0.366 1.185322e-20 Fibroblast ENSG00000074800</span></span>
<span id="cb10-33"><a href="#cb10-33" tabindex="-1"></a><span class="co">#&gt; ENSG00000112715 6.663398e-18 -1.1826785 0.078 0.202 1.549507e-13 Fibroblast ENSG00000112715</span></span>
<span id="cb10-34"><a href="#cb10-34" tabindex="-1"></a><span class="co">#&gt; ENSG00000140416 1.844156e-17 -0.6994000 0.175 0.326 4.288400e-13 Fibroblast ENSG00000140416</span></span>
<span id="cb10-35"><a href="#cb10-35" tabindex="-1"></a><span class="co">#&gt; ENSG00000125810 8.916133e-15  1.8102270 0.057 0.019 2.073358e-10 Fibroblast ENSG00000125810</span></span></code></pre></div>
</div>
<div id="store-the-results-in-lamindb" class="section level1">
<h1>Store the results in LaminDB</h1>
<p>Now that we have our results, we can save them to the LaminDB
instance.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>seu_path <span class="ot">&lt;-</span> <span class="fu">tempfile</span>(<span class="at">fileext =</span> <span class="st">&quot;.rds&quot;</span>)</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="fu">saveRDS</span>(seurat, seu_path)</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>db<span class="sc">$</span>Artifact<span class="sc">$</span><span class="fu">from_df</span>(</span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>  markers,</span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>  <span class="at">description =</span> <span class="st">&quot;Marker genes for renal cell carcinoma dataset&quot;</span></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a>)<span class="sc">$</span><span class="fu">save</span>()</span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a><span class="co">#&gt; [92m→[0m returning existing artifact with same hash: Artifact(uid=&#39;uo8EpZG3uVBHDAq60000&#39;, is_latest=True, description=&#39;Marker genes for renal cell carcinoma dataset&#39;, suffix=&#39;.parquet&#39;, type=&#39;dataset&#39;, size=11537, hash=&#39;N-D0s0VzXjS8IsIpuIF_Jw&#39;, _hash_type=&#39;md5&#39;, _accessor=&#39;DataFrame&#39;, visibility=1, _key_is_virtual=True, storage_id=2, transform_id=173, run_id=319, created_by_id=28, created_at=2024-11-20 10:19:51 UTC)</span></span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a></span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a>db<span class="sc">$</span>Artifact<span class="sc">$</span><span class="fu">from_path</span>(</span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a>  seu_path,</span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a>  <span class="at">description =</span> <span class="st">&quot;Seurat object for renal cell carcinoma dataset&quot;</span></span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a>)<span class="sc">$</span><span class="fu">save</span>()</span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a><span class="co">#&gt; ... uploading file60c2f2dc66740.rds:  0.0%... uploading file60c2f2dc66740.rds: 100.0%</span></span></code></pre></div>
</div>
<div id="close-the-connection" class="section level1">
<h1>Close the connection</h1>
<p>Finally, we can close the connection to the database.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a>db<span class="sc">$</span><span class="fu">finish</span>()</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="co">#&gt; [92m→[0m finished Run(&#39;CKjEjAr6&#39;) after 0d 0h 0m 35s at 2024-11-20 11:18:32 UTC</span></span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="co">#&gt; [92m→[0m go to: https://lamin.ai/laminlabs/lamindata/transform/I8BlHXFXqZOG0000</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a><span class="co">#&gt; [92m→[0m if you want to update your script without re-running it, use `lamin save script.py`</span></span></code></pre></div>
</div>
<div id="render-and-upload-the-notebook" class="section level1">
<h1>Render and upload the notebook</h1>
<p>You can render this notebook to HTML:</p>
<ul>
<li><p>In RStudio, click the “Knit” button</p></li>
<li><p>From the command line, run:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="ex">Rscript</span> <span class="at">-e</span> <span class="st">&#39;rmarkdown::render(&quot;example_workflow.Rmd&quot;)&#39;</span></span></code></pre></div></li>
<li><p>Or use the <code>rmarkdown</code> package in R:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a>rmarkdown<span class="sc">::</span><span class="fu">render</span>(<span class="st">&quot;example_workflow.Rmd&quot;</span>)</span></code></pre></div></li>
</ul>
<p>And then save it to your LaminDB instance using the
<code>lamin</code> CLI:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="ex">lamin</span> save example_workflow.Rmd</span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
