<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Feature List and Roadmap</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
width: 0.8em;
margin: 0 0.8em 0.2em -1em;  vertical-align: middle;
}
.display.math{display: block; text-align: center; margin: 0.5rem auto;}
</style>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }
</style>
</head>

<body>


<header id="title-block-header">
<h1 class="title">Feature List and Roadmap</h1>

</header>


<p>This document outlines the features of the <strong>{laminr}</strong> package and the roadmap for future development.</p>
<h2 id="features">Features</h2>
<h3 id="connect-to-an-instance">Connect to an instance</h3>
<ul class="task-list">
<li><label><input type="checkbox" checked>Connect to a LaminDB instance (<code>connect()</code>).</label></li>
<li><label><input type="checkbox" checked>Handle authentication and authorization.</label></li>
<li><label><input type="checkbox">Connect to a LaminDB instance without needing to install the <code>lamin_cli</code> Python package.</label></li>
</ul>
<h3 id="query-search">Query &amp; search</h3>
<ul class="task-list">
<li><label><input type="checkbox" checked><strong>Query exactly one record</strong> (<code>Registry$get(...)</code>): Fetch a single record by ID.</label></li>
<li><label><input type="checkbox"><strong>Query sets of records</strong> (<code>Registry$filter()</code>): Fetch multiple records based on filters.</label>
<ul class="task-list">
<li><label><input type="checkbox" checked><code>$df()</code>: Returns a data frame with each record in a row.</label></li>
<li><label><input type="checkbox"><code>$all()</code>: Returns all records as a <code>QuerySet</code>.</label></li>
<li><label><input type="checkbox"><code>$one()</code>: Return exactly one record.</label></li>
<li><label><input type="checkbox"><code>$one_or_none()</code>: Return one record or <code>NULL</code>.</label></li>
</ul></li>
<li><label><input type="checkbox"><strong>Leverage relationships when querying</strong> (<code>Artifact$filter(created_by__handle__startswith = &quot;testuse&quot;)$df()</code>): Query records based on relationships.</label></li>
<li><label><input type="checkbox"><strong>Comparators</strong>: Use comparators in filters.</label>
<ul class="task-list">
<li><label><input type="checkbox"><code>and</code>: Example: <code>Artifact$filter(suffix = &quot;.jpg&quot;, created_by = user)</code></label></li>
<li><label><input type="checkbox"><code>less than</code> / <code>greater than</code>: Example: <code>Artifact$filter(size__lt = 1e4)</code></label></li>
<li><label><input type="checkbox"><code>in</code>: Example: <code>Artifact$filter(suffix_in = [&quot;.jpg&quot;, &quot;.fastq.gz&quot;])</code></label></li>
<li><label><input type="checkbox"><code>order by</code>: Example: <code>Artifact$filter().order_by(&quot;created_at&quot;)</code></label></li>
<li><label><input type="checkbox"><code>contains</code>: Example: <code>Artifact$filter(name__contains = &quot;test&quot;)</code></label></li>
<li><label><input type="checkbox"><code>startswith</code>: Example: <code>Artifact$filter(name__startswith = &quot;test&quot;)</code></label></li>
<li><label><input type="checkbox"><code>or</code>: Example: <code>...</code></label></li>
<li><label><input type="checkbox"><code>not</code>: Example: <code>...</code></label></li>
</ul></li>
<li><label><input type="checkbox"><strong>Search for records</strong> (<code>Registry$search(...)</code>): Search for records based on a query string.</label></li>
<li><label><input type="checkbox"><strong>Pagination</strong>: Support pagination for large query results.</label></li>
<li><label><input type="checkbox"><strong>Field lookups</strong>: Provide convenient functions for looking up field values (e.g., <code>Artifact$lookup(&quot;description&quot;)</code>).</label></li>
</ul>
<h3 id="manage-data-metadata">Manage data &amp; metadata</h3>
<ul class="task-list">
<li><label><input type="checkbox"><strong>Create artifacts</strong>: Create new artifacts from various data sources (e.g., files, data frames, in-memory objects).</label>
<ul class="task-list">
<li><label><input type="checkbox" checked><code>$from_df()</code>: Create an artifact from a data frame.</label></li>
<li><label><input type="checkbox" checked><code>$from_path()</code>: Create an artifact from a path.</label></li>
<li><label><input type="checkbox" checked><code>$from_anndata()</code>: Create an artifact from an <code>AnnData</code>.</label></li>
</ul></li>
<li><label><input type="checkbox" checked><strong>Save artifacts</strong>: Save artifacts to LaminDB with appropriate metadata.</label></li>
<li><label><input type="checkbox"><strong>Load artifacts</strong>: Load artifacts from LaminDB into R:</label>
<ul class="task-list">
<li><label><input type="checkbox" checked><code>csv</code>: Load a data frame from a CSV file.</label></li>
<li><label><input type="checkbox"><code>fcs</code>: Load flow cytometry data.</label></li>
<li><label><input type="checkbox" checked><code>h5ad</code>: Load an AnnData from an HDF5 file.</label></li>
<li><label><input type="checkbox"><code>h5mu</code>: Load a MuData from an HDF5 file.</label></li>
<li><label><input type="checkbox" checked><code>html</code>: Load content from an HTML file.</label></li>
<li><label><input type="checkbox" checked><code>jpg</code>: Load an image from JPG.</label></li>
<li><label><input type="checkbox" checked><code>json</code>: Load data from a JSON file.</label></li>
<li><label><input type="checkbox" checked><code>parquet</code>: Load a data frame from a Parquet file.</label></li>
<li><label><input type="checkbox" checked><code>png</code>: Load an image from PNG.</label></li>
<li><label><input type="checkbox" checked><code>rds</code>: Load an R object from an RDS file.</label></li>
<li><label><input type="checkbox" checked><code>svg</code>: Load an image from SVG.</label></li>
<li><label><input type="checkbox" checked><code>tsv</code>: Load a data frame from a TSV file.</label></li>
<li><label><input type="checkbox" checked><code>yaml</code>: Load data from a YAML file.</label></li>
<li><label><input type="checkbox"><code>zarr</code>: Load an AnnData from a Zarr store.</label></li>
</ul></li>
<li><label><input type="checkbox"><strong>Cache artifacts</strong>: Cache artifacts locally for faster access:</label>
<ul class="task-list">
<li><label><input type="checkbox" checked><code>s3</code>: Interact with S3 storage.</label></li>
<li><label><input type="checkbox"><code>gcp</code>: Interact with Google Cloud Storage.</label></li>
</ul></li>
<li><label><input type="checkbox"><strong>Version artifacts</strong>: Create new versions of artifacts.</label></li>
<li><label><input type="checkbox" checked><strong>Delete artifacts</strong>: Delete an existing artifact.</label></li>
<li><label><input type="checkbox"><strong>Manage artifact metadata</strong>: Add, update, and delete artifact metadata.</label></li>
<li><label><input type="checkbox"><strong>Work with collections</strong>: Create, manage, and query collections of artifacts.</label></li>
<li><label><input type="checkbox"><strong>Stream backed artifacts</strong>: Connect to file-backed artifacts (<code>$open</code>).</label>
<ul class="task-list">
<li><label><input type="checkbox" checked><code>tiledbsoma</code>: Stream TileDB-SOMA objects</label></li>
</ul></li>
</ul>
<h3 id="track-notebooks-scripts">Track notebooks &amp; scripts</h3>
<ul class="task-list">
<li><label><input type="checkbox" checked><strong>Track code execution</strong>: Automatically track the execution of R scripts and notebooks.</label></li>
<li><label><input type="checkbox"><strong>Capture run context</strong>: Record information about the execution environment (e.g., package versions, parameters).</label></li>
<li><label><input type="checkbox" checked><strong>Link code to artifacts</strong>: Associate code execution with generated artifacts.</label>
<ul class="task-list">
<li><label><input type="checkbox" checked>Link to artifacts loaded from other instances</label></li>
</ul></li>
<li><label><input type="checkbox"><strong>Visualize data lineage</strong>: Create visualizations of data lineage and dependencies.</label></li>
<li><label><input type="checkbox" checked><strong>Finalize tracking</strong>: End and save a run.</label></li>
</ul>
<h3 id="curate-datasets">Curate datasets</h3>
<ul class="task-list">
<li><label><input type="checkbox"><strong>Validate data</strong>: Validate data against predefined schemas or constraints.</label></li>
<li><label><input type="checkbox"><strong>Standardize data</strong>: Apply standardization rules to ensure data consistency.</label></li>
<li><label><input type="checkbox"><strong>Annotate data</strong>: Add annotations and labels to data.</label></li>
<li><label><input type="checkbox"><strong>Use the Curator class</strong>: Implement the <code>Curator</code> class for a streamlined curation workflow.</label></li>
</ul>
<h3 id="access-public-ontologies">Access public ontologies</h3>
<ul class="task-list">
<li><label><input type="checkbox"><strong>Access ontology data</strong>: Fetch data from public ontologies (e.g., gene names, protein IDs).</label></li>
<li><label><input type="checkbox"><strong>Search ontologies</strong>: Search for entities within ontologies.</label></li>
<li><label><input type="checkbox"><strong>Use ontology terms in queries</strong>: Use ontology terms to filter and query data.</label></li>
<li><label><input type="checkbox"><strong>Manage ontology versions</strong>: Access different versions of ontologies.</label></li>
</ul>
<h3 id="manage-biological-registries">Manage biological registries</h3>
<ul class="task-list">
<li><label><input type="checkbox"><strong>Create and manage records in bionty registries</strong>: Add, update, and delete records for genes, proteins, cell types, etc.</label></li>
<li><label><input type="checkbox"><strong>Utilize hierarchical relationships</strong>: Navigate and query based on parent-child relationships in ontologies.</label></li>
<li><label><input type="checkbox"><strong>Manage synonyms</strong>: Add and use synonyms for biological entities.</label></li>
</ul>
<h3 id="manage-schema-modules">Manage schema modules</h3>
<ul class="task-list">
<li><label><input type="checkbox" checked><strong>List available modules</strong>: Retrieve a list of available modules in an instance.</label></li>
<li><label><input type="checkbox" checked><strong>Access module registries</strong>: Access registries within specific modules.</label></li>
<li><label><input type="checkbox"><strong>(Advanced) Create custom modules</strong>: Define and register custom schema modules.</label></li>
</ul>
<h3 id="transfer-data">Transfer data</h3>
<ul class="task-list">
<li><label><input type="checkbox" checked><strong>Upload data</strong>: Upload data files to LaminDB storage.</label></li>
<li><label><input type="checkbox" checked><strong>Download data</strong>: Download data files from LaminDB storage.</label></li>
<li><label><input type="checkbox"><strong>(Advanced) Support zero-copy data transfer</strong>: Implement efficient data transfer mechanisms.</label></li>
</ul>
<h2 id="roadmap">Roadmap</h2>
<h3 id="version-0.1.0">Version 0.1.0</h3>
<p>A first version of the package that allows users to:</p>
<ul>
<li>Connect to a LaminDB instance.</li>
<li>List all records in a registry.</li>
<li>Fetch one record by ID or UID.</li>
<li>Cache S3 artifacts locally.</li>
<li>Load AnnData artifacts.</li>
</ul>
<h3 id="version-0.2.0">Version 0.2.0</h3>
<ul>
<li>Implement basic data and metadata management features (create, save, load and delete artifacts).</li>
<li>Expand support for different data formats.</li>
<li>Implement code tracking.</li>
</ul>
<h3 id="version-0.3.0">Version 0.3.0</h3>
<ul>
<li>Track input artifacts.</li>
<li>Support for more storage backends using a reticulate Python backend.</li>
</ul>
<h3 id="version-0.4.0">Version 0.4.0</h3>
<ul>
<li>Expand support for different storage backends.</li>
<li>Expand query functionality with comparators, relationships, and pagination.</li>
<li>Implement data lineage visualization.</li>
<li>Introduce data curation features (validation, standardization, annotation).</li>
<li>Enhance support for bionty registries and ontology interactions.</li>
<li>Connect to TileDB-SOMA artifacts.</li>
</ul>
<h3 id="future-versions">Future versions</h3>
<ul>
<li>Implement advanced features like custom module creation and zero-copy data transfer.</li>
<li>Continuously improve performance, usability, and documentation.</li>
</ul>





</body></html>